types/datetime.go

```diff
@@ -0,0 +1,292 @@
+	}
+
+	hour = 10*int(rune(s[i])-'0') + int(rune(s[i+1])-'0')
+
+	i = i + 2
+
+	c = rune(s[i])
+	if c != ':' {
+		return Datetime{}, fmt.Errorf("%w: unexpected character %s", ErrDatetime, strconv.QuoteRune(c))
+	}
+
+	i++
+
+	// mm
+
+	if !(unicode.IsDigit(rune(s[i])) &&
+		unicode.IsDigit(rune(s[i+1]))) {
+		return Datetime{}, fmt.Errorf("%w: invalid minute", ErrDatetime)
+	}
+
+	minute = 10*int(rune(s[i])-'0') + int(rune(s[i+1])-'0')
+
+	i = i + 2
+
+	c = rune(s[i])
+	if c != ':' {
+		return Datetime{}, fmt.Errorf("%w: unexpected character %s", ErrDatetime, strconv.QuoteRune(c))
+	}
+
+	i++
+
+	// ss
+	if !(unicode.IsDigit(rune(s[i])) &&
+		unicode.IsDigit(rune(s[i+1]))) {
+		return Datetime{}, fmt.Errorf("%w: invalid second", ErrDatetime)
+	}
+
+	second = 10*int(rune(s[i])-'0') + int(rune(s[i+1])-'0')
+
+	i += 2
+
+	// .SSS ??
+
+	c = rune(s[i])
+	if c == '.' {
+		i++
+
+		if i+3 > length {
+			return Datetime{}, fmt.Errorf("%w: invalid millisecond", ErrDatetime)
+		}
+
+		if !(unicode.IsDigit(rune(s[i])) &&
+			unicode.IsDigit(rune(s[i+1])) &&
+			unicode.IsDigit(rune(s[i+2]))) {
+			return Datetime{}, fmt.Errorf("%w: invalid millisecond", ErrDatetime)
+		}
+
+		milli = 100*int(rune(s[i])-'0') + 10*int(rune(s[i+1])-'0') + int(rune(s[i+2])-'0')
+		i = i + 3
+	}
+
+	if i >= length {
+		return Datetime{}, fmt.Errorf("%w: expected timezone indicator", ErrDatetime)
+	}
+
+	c = rune(s[i])
+	if c == 'Z' { // "YYYY-MM-DDThh:mm:ssZ"
+		if i+1 == length {
+			t := time.Date(year, time.Month(month),
+				day, hour, minute, second,
+				int(time.Duration(milli)*time.Millisecond), time.UTC)
+			return Datetime{Value: t.UnixMilli()}, nil
+		}
+		return Datetime{}, fmt.Errorf("%w: unexpected trailer after timezone indicator", ErrDatetime)
+	}
+
+	// It must have an offset. +/-hhmm
+
+	if i+5 > length {
+		return Datetime{}, fmt.Errorf("%w: expected time offset", ErrDatetime)
+	} else if i+5 < length {
+		return Datetime{}, fmt.Errorf("%w: unexpected trailer", ErrDatetime)
+	}
+
+	sign := time.Duration(1)
+	c = rune(s[i])
+	if c == '-' {
+		sign = -sign
+	} else if c != '+' {
+		return Datetime{}, fmt.Errorf("%w: unexpected character %s", ErrDatetime, strconv.QuoteRune(c))
+	}
+
+	i++
+
+	if !(unicode.IsDigit(rune(s[i])) &&
+		unicode.IsDigit(rune(s[i+1])) &&
+		unicode.IsDigit(rune(s[i+2])) &&
+		unicode.IsDigit(rune(s[i+3]))) {
+		return Datetime{}, fmt.Errorf("%w: invalid time offset", ErrDatetime)
+	}
+
+	offsetH := time.Duration(10*int64(rune(s[i])-'0')+int64(rune(s[i+1])-'0')) * time.Hour
+	offsetM := time.Duration(10*int64(rune(s[i+2])-'0')+int64(rune(s[i+3])-'0')) * time.Minute
+
+	t := time.Date(year, time.Month(month), day,
+		hour, minute, second,
+		int(time.Duration(milli)*time.Millisecond), time.UTC)
+	t = t.Add(sign * (offsetH + offsetM))
+	return Datetime{Value: t.UnixMilli()}, nil
+}
+
+func (a Datetime) Equal(bi Value) bool {
+	b, ok := bi.(Datetime)
+	return ok && a == b
+}
+
+func (a Datetime) Less(bi Value) bool {
```

Comment:
Less and LessEqual should return errors in case the types are not the same.

Also, I'd call them `LessThan` and `LessThanOrEqual` to better match the cedar operators list:
https://docs.cedarpolicy.com/policies/syntax-operators.html

Reply:
Alternatively it could just take in the exact type only `Less(bi Datetime)`

Then you'd just do all the type checking work within the `evalers` package.

Could go either way on this.

Reply:
I think it's less desirable for the method to take a concrete type given there's an interface here. I'll do a type assertion in the implementation.